examples/bmi-for-nextgen.ipynb

```diff
@@ -0,0 +1,1696 @@
+    "\n",
+    "The other thing this particular `initialize` function does is map the CSDMS standard names to the model variable names. We also give the UDUNITS-compliant units for each input and output variable. \n",
+    "\n",
+    "``` python\n",
+    "def initialize(self, filename=None):\n",
+    "        \"\"\"Initialize the model.\n",
+    "\n",
+    "        Parameters\n",
+    "        ----------\n",
+    "        filename : str, optional\n",
+    "            Path to name of input file.\n",
+    "        \"\"\"\n",
+    "        if filename is None:\n",
+    "            self._model = Snow()\n",
+    "        elif isinstance(filename, str):\n",
+    "            with open(filename, \"r\") as file_obj:\n",
+    "                self._model = Snow.from_file_like(file_obj.read())\n",
+    "        else:\n",
+    "            self._model = Snow.from_file_like(filename)\n",
+    "\n",
+    "        self._values = {\"atmosphere_water__precipitation_leq-volume_flux\": self._model.ppt_mm,\n",
+    "                        \"land_surface_air__temperature\": self._model.tair_c,\n",
+    "                        \"snowpack__liquid-equivalent_depth\": self._model.swe_mm,\n",
+    "                        \"snowpack__melt_volume_flux\": self._model.melt_mm}\n",
+    "        self._var_units = {\"atmosphere_water__precipitation_leq-volume_flux\": \"mm d-1\",\n",
+    "                           \"land_surface_air__temperature\": \"C\",\n",
+    "                           \"snowpack__liquid-equivalent_depth\": \"mm\",\n",
+    "                           \"snowpack__melt_volume_flux\": \"mm d-1\"}\n",
+    "```\n",
+    "\n",
+    "### 3.3.2 BMI::update()\n",
+    "\n",
+    "Next we have a very short `update` function that calls `advance_in_time` which moves the model one time step and runs the `solve_snow` function we saw earlier. This is an example of a BMI function calling model functions that already existed.\n",
+    "\n",
+    "*You should not copy and paste all of your model's code into the `update` function.* Yes, you can technically do that, but just don't.\n",
+    "\n",
+    "``` python\n",
+    "def update(self):\n",
+    "        \"\"\"Advance model by one time step.\"\"\"\n",
+    "        self._model.advance_in_time()\n",
+    "```\n",
+    "### 3.3.3 BMI::finalize()\n",
+    "\n",
+    "This is a simple function to implement for `snowBMI`, but lower-level languages with greater memory-handling challenges (I'm looking at you, C...) will be more complex.\n",
+    "\n",
+    "```python\n",
+    "    def finalize(self):\n",
+    "        \"\"\"Finalize model.\"\"\"\n",
+    "        self._model = None\n",
+    "```\n",
+    "\n",
+    "### 3.3.4 BMI::get_value() and BMI::set_value()\n",
```

Comment:
I think it is important here to emphasize that nextgen compatible BMI modules must implement get_var_nbytes and get_var_itemsize and get_var_type correctly ALONG with get/set value in order to provide the language interoperability it does.  Don't need to cover them in detail, but they deserve a shout out as "must not be BMI_FAILURE" if you want to this to work with ngen

Reply:
Good catch. I've added that info.